### ISSUE TITLE
PINN does not boot from usb

### ISSUE THREAD

**Issue by swekley** [procount/pinn#197]
PINN works when i copy the files on a fat32 formatted sd card, but I get the splash screen after 2 minutes when trying it with my SanDisk Extreme Pro usb stick.

I'm talking about this one: https://www.sandisk.com/home/usb-flash/extremepro-usb

Raspbian runs fine from it, however I don't want to stick with just raspbian on a 128gb stick and I'd prefer not to use any sd cards

I'm using a rpi 3b with the usb boot bit set

**Comment by procount** [comment]
> but I get the splash screen after 2 minutes when trying it with my SanDisk Extreme Pro usb stick.

So, do you mean that PINN is not booting from your usb stick and all you get is the splash screen, even after 2 minutes from powering on?

If Raspbian will boot from your usb stick, then there is no reason for PINN not to. However, it is possible you are having problems formatting your USB stick correctly for PINN as it requires FAT16 or FAT32 file format, which is a bit harder to achieve on a 128GB USB stick as the default for that is exFAT. Are you getting any LEDs flashing in a regular pattern to indicate an error message?

I suggest you download my Etcher image version of PINN from https://sourceforge.net/projects/pinn/files/Etcher/pinn-lite.img.zip/download and flash it to your USB stick using Etcher (No need to unzip it first as Etcher will take care of that) and report back whether that fixes it.

**Comment by swekley** [comment]
Yes, well the screen stays black for 2 minutes and then it gives the splash screen. I'll try the Etcher image. USB is correctly formatted as fat32 using osx terminal's diskutil

**Comment by swekley** [comment]
Using the Etcher image it's working fine now.
Thanks!

**Comment by lurch** [comment]
> which is a bit harder to achieve on a 128GB USB stick as the default for that is exFAT

See https://www.raspberrypi.org/documentation/installation/sdxc_formatting.md :slightly_smiling_face:

**Comment by procount** [comment]
> Yes, well the screen stays black for 2 minutes and then it gives the splash screen

That's a really slow startup. I hope it is much better than that now you have it working with my Etcher image.
USB booting is slower, but I believe you can speed it up by leaving a small SD card fitted with just bootcode.bin on it (as you would to get USB boot on a Rpi 0/1/2.

**Comment by lurch** [comment]
ping @XECDesign and @ghollingworth
Does it seem plausible that a USB-boot would take two minutes just to get to the splash screen?

**Comment by XECDesign** [comment]
I don't know if 2 minutes is right, but it would take a while. Remember the bootrom has fairly minimal code to enumerate USB devices, find and load start.elf block by block (which is relatively large). It's certainly not optimised for speed. Then add the time it takes for it to stop trying to boot from the SD card.

**Comment by ghollingworth** [comment]
It shouldn't take that long, but it's possible it's doing something else (i.e. waiting for HDMI for example).  You could enable BOOT_UART and look at the UART output to see what is happening...

https://www.raspberrypi.org/documentation/hardware/raspberrypi/bootmodes/

**Comment by procount** [comment]
For comparison, I flashed PINN to my Sandisk Ultra USB 3.0 16GB USB Stick with Etcher
I booted it on my Pi 3B with no SD card fitted. 

10 secs to rainbow splash screen.
29 secs til the splash screen disappeared and the screen went black
34 secs for PINN's Raspberry logo to display & card formatting to start.
43 secs for PINN's recovery menu to come up and start scanning for OSes

I then re-formatted the whole stick as FAT 32 with win32diskimager and copied the PINN files across manually and timed the boot.

10 secs to rainbow splash screen.
26 secs til the splash screen disappeared and the screen went black
31 secs for PINN's Raspberry logo to display & card formatting to start.
39 secs for PINN's recovery menu to come up and start scanning for OSes

Timing was only by my watch, but nowhere near 2 minutes,  

EDIT: Using a Sandisk Ultra SDHC I class 10 16GB SD card with PINN written by Etcher:
4s/13s/17s/26s (I was a bit slow starting my stopwatch, so maybe -1s off each)
